lib/logstash/codecs/csv.rb

```diff
@@ -108,10 +113,41 @@ def register
     @logger.debug? && @logger.debug("CSV parsing options", :col_sep => @separator, :quote_char => @quote_char)
   end
 
-  def decode(data)
-    data = @converter.convert(data)
+  def decode(data, &block)
+    puts("*** data=#{data.inspect}")
+    @buffer.extract(data).each do |line|
+      puts("*** line=#{line.inspect}")
+      parse(@converter.convert(line), &block)
+    end
+  end
+
+  def encode(event)
+    if @include_headers
+      csv_data = CSV.generate_line(select_keys(event), :col_sep => @separator, :quote_char => @quote_char, :headers => true)
+      @on_event.call(event, csv_data)
+
+      # output headers only once per codec lifecycle
+      @include_headers = false
+    end
+
+    csv_data = CSV.generate_line(select_values(event), :col_sep => @separator, :quote_char => @quote_char)
+    @on_event.call(event, csv_data)
+  end
+
+  def flush(&block)
+    remainder = @buffer.flush
+    if !remainder.empty?
+      puts("flush: remainder=#{remainder}")
+      parse(@converter.convert(remainder), &block)
+    end
+  end
+
+  private
+
+  def parse(line, &block)
     begin
-      values = CSV.parse_line(data, :col_sep => @separator, :quote_char => @quote_char)
+      values = CSV.parse_line(line, :col_sep => @separator, :quote_char => @quote_char)
```

Comment:
What about something like:

~~~ ruby
CSV.parse(line, :col_sep => @separator, :quote_char => @quote_char).each do |values|
  next if values.nil?

  ## same implementation, using `next` instead of `return` when finding column names
end
~~~